### ISSUE TITLE
ImageNet Training Support

### ISSUE THREAD

**Issue by YukeWang96** [Xilinx/brevitas#243]
Hi, All

Is there any suggestion for using this for ImageNet training? 

I only found the evaluation script on ImageNet at `https://github.com/Xilinx/brevitas/tree/master/brevitas_examples/imagenet_classification`.

Thanks.

**Comment by volcacius** [comment]
Hello,

For the most part there isn't really anything special about training for ImageNet. Once you have your quantized model defined you can just throw it inside your training loop and that's it really. 
Depending on whether you are retraining from a floating point model or training from scratch you should adjust your hyperparemeters accordingly (e.g. if you are retraining from float to 8b you probably want a relatively low learning rate), but in general in my experience between 3 and 8 bits the way learning goes is quite similar to how it would go if you were training a float model. You get more fiddly behaviour at 1 and 2 bits.
Last thing (not specific to ImageNet, but useful to know), if you are retraining from a floating point model you will probably get errors around missing state dict keys. This is because Brevitas is learning a few extra nn.Parameters on top of the original model. In that case just launch training with the env variable BREVITAS_IGNORE_MISSING_KEYS=1.

Alessandro

**Comment by YukeWang96** [comment]
> Hello,
> 
> For the most part there isn't really anything special about training for ImageNet. Once you have your quantized model defined you can just throw it inside your training loop and that's it really.
> Depending on whether you are retraining from a floating point model or training from scratch you should adjust your hyperparemeters accordingly (e.g. if you are retraining from float to 8b you probably want a relatively low learning rate), but in general in my experience between 3 and 8 bits the way learning goes is quite similar to how it would go if you were training a float model. You get more fiddly behaviour at 1 and 2 bits.
> Last thing (not specific to ImageNet, but useful to know), if you are retraining from a floating point model you will probably get errors around missing state dict keys. This is because Brevitas is learning a few extra nn.Parameters on top of the original model. In that case just launch training with the env variable BREVITAS_IGNORE_MISSING_KEYS=1.
> 
> Alessandro

I also observe that after each step of training, there is something like` weight.clip(-1.1)`, is that necessary for all datasets and models with quantization?

Thanks

**Comment by volcacius** [comment]
Hello,

The weight clipping is kind of specific to when you have constant scaling (https://github.com/Xilinx/brevitas/blob/30da8ab5376af7dac814b0c69d881f8f3b92255c/brevitas_examples/bnn_pynq/models/common.py#L34), which is typically used on smaller datasets like the ones in the bnn_pynq training example to help with regularization. 
For larger datasets like imagenet you typically want scaling based on statistics (https://github.com/Xilinx/brevitas/blob/30da8ab5376af7dac814b0c69d881f8f3b92255c/brevitas_examples/imagenet_classification/models/common.py#L19) and that doesn't require further clipping. 
Obviously none of this is an exact science and I encourage you to do your own experiments, but based on my experience (I trained a lot of quantized imagenet models) that's what I would do.

Alessandro

**Comment by YukeWang96** [comment]
I have another question is 

> Do we have to set the precision of the input layer and output layer (e.g., keep them even as full precision) higher than the layers in the middle?

Thanks!

**Comment by volcacius** [comment]
The first layer yes, I'd do 4b or higher, anything lower you are gonna notice the difference a lot. The last layer not necessarily, I've gone down to 2b weights in some cases with good accuracy, but it's gonna take longer to converge and you get more sensitive to how are you dealing with biases.   
In my experience, if I keep everything at 4b (weights and activations of all layers including first and last) I usually expect a good training run to get around 1.75%-2% absolute accuracy loss on ImageNet val compared to the baseline float accuracy on depthwise-heavy topologies like MobileNet/EfficientNet (1.5%-1.75% if I'm really pushing it, but it can be hard). Around 7% with everything at 3b.  
In general if you don't want to spend too much effort 4b is usually a good choice for modern ImageNet convnets.

Alessandro

**Comment by YukeWang96** [comment]
Thanks a lot!

When I am looking at the example in this repo, I find that in order for training Quantized NN. I need

> 1.  input quantization before the layer. (e.g, `QuantIdentity`)
> 2. set the layer with quantization bit. (e.g., `QuantConv2d(..., weight_bit_width=weight_bit_width)`)
> 3. quantization on output . (e.g., `QuantIdentity`)

is this a correct process for quantized training? or it is a little bit redundant?

**Comment by YukeWang96** [comment]
I also have another observation

> By loading the pre-trained model from PyTorch for imagNet, and replace the conv and linear with the quantized version in the repo, the training accuracy show very low which is almost similar to training from scratch.

Could you give me some suggestions on this?

**Comment by volcacius** [comment]
Hello,

> 
> 
> Thanks a lot!
> 
> When I am looking at the example in this repo, I find that in order for training Quantized NN. I need
> 
> > 1. input quantization before the layer. (e.g, `QuantIdentity`)
> > 2. set the layer with quantization bit. (e.g., `QuantConv2d(..., weight_bit_width=weight_bit_width)`)
> > 3. quantization on output . (e.g., `QuantIdentity`)
> 
> is this a correct process for quantized training? or it is a little bit redundant?
> One more thing I noticed is that when adding more `QuantIdentity` layers the training becomes very slow.

It's a way to approach it, but there is too much redundancy. Every output of a layer is the input of another layer, so you are quantizing everything twice, which contributes to make everything very slow (although a slow down is expected).
You should have a QuantIdentity to quantize the input to the first layer, QuantConv2d to replace Conv2d, and QuantReLU to replace ReLUs. That is going to be supported by out FINN backend too.

> 
> 
> I also have another observation
> 
> > By loading the pre-trained model from PyTorch for imagNet, and replace the conv and linear with the quantized version in the repo, the training accuracy show very low which is almost similar to training from scratch.
> 
> Could you give me some suggestions on this?

Regarding this, Brevitas doesn't do any form of advanced direct quantization, meaning the initialization of quantization is relatively simple. So at the beginning of training it usually looks like it lost all accuracy. At 8b it's gonna recover pretty quickly, at lower precision it's gonna take more time. In general at >= 3b it is helpful to retrain from the pretrained float models. At 1-2b it's typically better to train from scratch.
 
Alessandro

**Comment by YukeWang96** [comment]
How about when we build a network by using the `Batchnorm` and `avgPool` layer of Pytorch, should we consider adding the `quantidentity` after these layers?

Thanks!

**Comment by volcacius** [comment]
It really depends on specifics of the topology and how you are deploying your neural network. If you don't have an activation layer after batch norm you should add a QuantIdentity, say for example before a residual add in certain topologies. Regarding AvgPool, we have a QuantAvgPool that implements add and shift with truncation in place of add and divide in order to perform an averaging, which makes sense if your target hardware can do that. If you don't have a specific target hardware in mind, you can also just do a standard AvgPool + QuantIdentity.

Alessandro
